### PR TITLE
feat(ui5-barcode-scanner-dialog): Initial implementation

### PR DESCRIPTION
**Pull request by kineticjs**

### CODE REVIEW

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+		 * Fired when the scan completed successfuuly.
+		 *
+		 * @event sap.ui.webcomponents.fiori.BarcodeScannerDialog#scan-success
+		 * @param {String} result the scan result.
+		 * @public
+		 */
+		 "scan-success": {
+			detail: {
+				text: { type: String },
+				rawBytes: { type: Object },
```
Comment (ilhan007) [inline packages/fiori/src/BarcodeScannerDialog.js:65]: rawBytes not documented as a param

File: packages/fiori/src/themes/BarcodeScannerDialog.css [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+.ui5-barcode-scanner-dialog-video {
+    object-fit: cover;
+}
+
+.ui5-barcode-scanner-dialog-footer {
+    display: flex;
+    justify-content: flex-end;
+    width: 100%;
+    padding-right: 0.5rem;
```
Comment (ilhan007) [inline packages/fiori/src/themes/BarcodeScannerDialog.css:15]: check if in RTL you should mirror the padding with padding-left: 0.5rem

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+		},
+	},
+};
+
+/**
+ * @class
+ *
+ * <h3 class="comment-api-title">Overview</h3>
+ *
+ * Opening the dialog launches the device camera and scans for known barcode formats.
```
Comment (ilhan007) [inline packages/fiori/src/BarcodeScannerDialog.js:89]: Maybe adding one additional intro sentence before "Opening the dialog launches the device camera..." or perhaps you can check it with the team's UA.

File: packages/fiori/src/BarcodeScannerDialog.js (hunk 1 above)
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js:66]: Done

File: packages/fiori/src/BarcodeScannerDialog.js (hunk 3 above)
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js:91]: Done. Will send to UA as well.
(btw Is there a UA assigned for the ui5-web-components in particular)?
Thnx

File: packages/fiori/src/themes/BarcodeScannerDialog.css (hunk 2 above)
Comment (kineticjs) [reply packages/fiori/src/themes/BarcodeScannerDialog.css]: Done

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+ * The <code>BarcodeScannerDialog</code> provides barcode scanning functionality for all devices that support the <code>MediaDevices.getUserMedia()</code> native API.
+ * Opening the dialog launches the device camera and scans for known barcode formats.
+ * <br>
+ * <br>
+ * Fires a <code>scanSuccess</code> event whenever a barcode is identified
+ * and a <code>scanError</code> event when the scan failed (e.g. due to missing permisions).
+ * <br>
+ * <br>
+ * The component internally uses the <ui5-link target="_blank" href="https://github.com/zxing-js/library">zxing-js/library</ui5-link> third party OSS.
+ * Check the zxing-js/library documentation for the list of supported barcode formats.
```
Comment (xtatica) [inline packages/fiori/src/BarcodeScannerDialog.js:99]:
```suggestion
 * The <code>BarcodeScannerDialog</code> component provides barcode scanning functionality for all devices that
 * support the <code>MediaDevices.getUserMedia()</code> native API.
 * Opening the dialog launches the device camera and scans for known barcode formats.
 * <br>
 * <br>
 * A <code>scanSuccess</code> event fires whenever a barcode is identified
 * and a <code>scanError</code> event fires when the scan failed (for example, due to missing permisions).
 * <br>
 * <br>
 * Internally, the component uses the zxing-js/library third party OSS.
 * For a list of supported barcode formats, see the <ui5-link target="_blank" href="https://github.com/zxing-js/library">zxing-js/library</ui5-link> documentation.
```

File: packages/fiori/test/pages/BarcodeScannerDialog.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<meta http-equiv="X-UA-Compatible" content="IE=edge">
+	<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
+	<meta charset="utf-8">
+
+	<title>Barcode scanner</title>
```
Comment (xtatica) [inline packages/fiori/test/pages/BarcodeScannerDialog.html:8]:
```suggestion
	<title>Barcode Scanner</title>
```

File: packages/fiori/test/specs/BarcodeScannerDialog.spec.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+const assert = require("chai").assert;
+const PORT = require("./_port.js");
+
+
+describe("BarcodeScannerDialog Behavior", () => {
+	before(() => {
+		browser.url(`http://localhost:${PORT}/test-resources/pages/BarcodeScannerDialog.html`);
+	});
+
+	it("fires scan-error when no persions granted", () => {
```
Comment (xtatica) [inline packages/fiori/test/specs/BarcodeScannerDialog.spec.js:11]:
```suggestion
	it("fires scan-error when no permissions granted", () => {
```

**Review by xtatica — CHANGES_REQUESTED** [review]

**Comment by ilhan007** [comment]
You will have to adopt this [change](https://github.com/SAP/ui5-webcomponents/pull/3528) (Dialog's `open` method has been renamed to `show` as we plan to use "open"
for a property name to declaratively open/close the dialog) https://github.com/SAP/ui5-webcomponents/pull/3528

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+ * @public
+ */
+const metadata = {
+	tag: "ui5-barcode-scanner-dialog",
+	languageAware: true,
+	slots: /** @lends  sap.ui.webcomponents.fiori.BarcodeScannerDialog.prototype */ {
+	},
+	properties: /** @lends  sap.ui.webcomponents.fiori.BarcodeScannerDialog.prototype */ {
+		/**
+		 * Indicates whether a loading indicator should be shown in the dialog.
```
[the remark this answers is not in the document]
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js]: Done

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+		 * @defaultvalue false
+		 * @private
+		 */
+		loading: {
+			type: Boolean,
+		},
+	},
+	events: /** @lends  sap.ui.webcomponents.fiori.BarcodeScannerDialog.prototype */ {
+		/**
+		 * Fired when the scan completed successfuuly.
```
[the remark this answers is not in the document]
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js]: Done

File: packages/fiori/src/BarcodeScannerDialog.js [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+		 *
+		 * @event sap.ui.webcomponents.fiori.BarcodeScannerDialog#scan-success
+		 * @param {String} text the scan result as string.
+		 * @param {Object} rawBytes the scan result as a Uint8Array.
```
[the remark this answers is not in the document]
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js]: Done

File: packages/fiori/src/BarcodeScannerDialog.js (hunk 4 above)
Comment (kineticjs) [reply packages/fiori/src/BarcodeScannerDialog.js]: Done

File: packages/fiori/test/pages/BarcodeScannerDialog.html (hunk 5 above)
Comment (kineticjs) [reply packages/fiori/test/pages/BarcodeScannerDialog.html]: Done

File: packages/fiori/test/specs/BarcodeScannerDialog.spec.js (hunk 6 above)
Comment (kineticjs) [reply packages/fiori/test/specs/BarcodeScannerDialog.spec.js]: Done

**Comment by dobrinyonkov** [comment]
It looks good to me! :)

**Review by ilhan007 — APPROVED** [review]

**Comment by kineticjs** [comment]
@ilhan007 ,

Thanks for the review on Friday! After retesting everything before merge, I found I has forgotten to adopt the "show" function in the playground sample, do made another update.
Also, I added an check for "loading" state in the beginning of the "show" function of the dialog.

Sorry for making these changes only now... I retested the functionality and the pages. The change is again for review.

Thanks!

**Review by ilhan007 — APPROVED** [review]

**Comment by ilhan007** [comment]
Looks good to me, someone from @SAP/ui5-webcomponents-topic-p needs to approve the change as well.

**Review by dobrinyonkov — APPROVED** [review]

**Pull request merged**